add document elaborating on how to use TLS and examples (#391)
kubernetes-sigs/service-apis: docs-src/tls.md

# TLS

Service APIs allow for a variety of ways to configure TLS. This document lays
out various TLS settings and gives general guidelines on how to use them
effectively.

## Client/Server and TLS

![overview](tls-overview.svg)

For Gateways, there are two connections involved:

- **downstream**: This is the connection between the client and the Gateway.
- **upstream**: This is the connection between the Gateway and backend resources
   specified by routes. These backend resources will usually be Services.
  
With Service APIs, TLS configuration of downstream and
upstream connections is managed independently.

Please note that in case of `Passthrough` TLS mode, no TLS settings take
effect as the TLS session from the client is NOT terminated at the Gateway.

> The rest of the document assumes that TLS is being terminated at the Gateway.

Reviewer: Perhaps state that "Terminate" is the default.  I noticed that some examples specify `mode: Terminate` and some omit it (but get `Terminate` because it is the default).

## Downstream TLS

Downstream TLS settings are configured using Listeners at the Gateway level.

### Listener and TLS 

Listeners expose the TLS setting on a per domain or sub-domain basis.
TLS settings of a listener are applied to all domains that satisfy the
`hostname.match` criteria.

In the following example, a Gateway is configured to serve a TLS certificate

> for any request that is served a Gateway.

Reviewer: ```suggestion
for any request that is served by the Gateway.
```

> for any request that is served a Gateway.

Reviewer 2: Either drop ` that is served a Gateway` or reword.

```yaml
listeners:

> - protocol: HTTPS # Other possible value is `TLS`

Reviewer 2: This section is titled "Listener and TLS" but the protocol is `HTTPS` with "# Other possible value is `TLS`. Either rename this section to "Listener and HTTPS" or provide background of the `protocol` field before diving into this example.

  port: 443
  hostname:
    match: Any # Other possible values are `Domain` and `Exact`
  tls:
    mode: Terminate
    certificateRef:
      kind: Secret
      group: core
      name: default-cert
    routeOverride:
      certificate: Deny
```

If `hostname.match` is set to `Exact`, then the TLS settings apply to only the
specific hostname that is set in `hostname.name`. This is the recommended way
of configuring TLS settings in Service APIs as it does not involve any

> redirection and should suffice in the majority of use-cases.

Reviewer: What does "redirection" refer to here?

Author: Redirection in the sense that TLS configuration is directly defined in the Gateway and doesn't require a Cluster Operator to check all routes for TLS definitions.

Reviewer: Ah, as in cross-referencing resources?  And does "This" in "This is the recommended way" refer to setting `tls.routeOverride.certificate: Deny`?  So if I understand correctly, this statement could be rephrased as, "Specifying `tls.routeOverride.certificate: Deny` is recommended because it centralizes TLS configuration within the Gateway specification and should suffice for the majority of use-cases."

Author: Updated.

> Please take a look at the examples below for various possibilities.

Reviewer: ```suggestion
Please take a look at the examples below for various alternatives.
```

> Please take a look at the examples below for various possibilities.

Reviewer 2: s/take a look at/review/

### Route and TLS

> If `listener.tls.routeOverride.certificate` is set to `Allow`, TLS certificates

Reviewer: ```suggestion
If `listeners[].tls.routeOverride.certificate` is set to `Allow`, TLS certificates
```

> can be configured at the Route-level. This feature is primarily meant for a

Reviewer 2: s/at the Route-level/on Routes that are bound to the Gateway/

Reviewer 2: Although someone can infer `routeOverride` will override the gateway listener cert with the route cert, it might be a good idea to explicitly call this out here.

cluster with a self-service model where Service owners bring their own TLS
certificates. This feature also mirrors the behavior of TLS as defined in the
Ingress v1 resource. One should use this feature only when the Gateway Admin

> wishes to delegate TLS to Service or App owners.

Reviewer: I don't think "Gateway Admin" and "App owner" are defined anywhere.  Is defining "Gateway Admin" covered by https://github.com/kubernetes-sigs/service-apis/issues/95?  I guess we can come back and update this text if needed depending on what role names are defined in https://github.com/kubernetes-sigs/service-apis/issues/95.

When using this feature, please note that the TLS certificate to serve is chosen

> before an HTTPRoute is selected. This is because the TLS handshake is completed

Reviewer 2: I'm a bit confused by "When using this feature, please note that the TLS certificate to serve is chosen
before an HTTPRoute is selected." If `listener.tls.routeOverride.certificate` is set to `Allow`, then the cert is configured at the route-level.

Reviewer 2: Reading further, I think I have a better understanding of what the above sentence is trying to convey. It might be helpful to introduce the `BackendPolicy` here for clarification.

> before an HTTP request is sent from the client.

Reviewer: This makes me wonder:  Does the same restriction apply if `spec.listeners[].routes.kind` is set to `TLSRoute`?  For HTTPRoutes, the restriction derives directly from the interaction between TLS and HTTP, but TLSRoutes aren't subject to the same limitation.  If this hasn't been decided already, I would suggest that we apply the same restriction to TLSRoutes (and perhaps document it here), even if it isn't strictly necessary, in order to provide a consistent API and clear semantics for implementers.

Author: It does apply to TLSRoute. Is this the right place to document it though? Should it not be in API docs rather?

> [TLS Certificate in Route](#tls-certificate-in-route) provides an example
> on how this feature can be used.

Reviewer: ```suggestion
[TLS Certificate in Route](#tls-certificate-in-route) provides an example
of how this feature can be used.
```
Using "of" feels more correct to me, but this might be a regional difference.

### Examples

#### TLS in listener

> In this example, Gateway is configured to serve `foo.example.com` and
> `bar.example.com` domains. The certificate for these domains are specified
> in Gateway.

Reviewer: ```suggestion
In this example, the Gateway is configured to serve the `foo.example.com` and
`bar.example.com` domains. The certificate for these domains is specified
in the Gateway.
```

```
{% include 'tls-basic.yaml' %}
```

#### Wildcard TLS listeners

> In this example, the Gateway is configured with wildcard certificate for
> `*.example.com` and a different certificate for `foo.example.com`.
> Since a specific match takes priority, the Gateway will serve the
> `foo-example-com-cert` for requests to `foo.example.com` and the
> `wildcard-example-com-cert` for all other requests.

Reviewer: ```suggestion
In this example, the Gateway is configured with a wildcard certificate for
`*.example.com` and a different certificate for `foo.example.com`.
Since a specific match takes priority, the Gateway will serve
`foo-example-com-cert` for requests to `foo.example.com` and
`wildcard-example-com-cert` for all other requests.
```
or
```suggestion
In this example, the Gateway is configured with a wildcard certificate for
`*.example.com` and a different certificate for `foo.example.com`.
Since a specific match takes priority, the Gateway will serve
the certificate in `foo-example-com-cert` for requests to `foo.example.com` and
the certificate in `wildcard-example-com-cert` for all other requests.
```

```yaml
{% include 'wildcard-tls-gateway.yaml' %}
```

#### TLS Certificate in Route

> In this example, Gateway is configured with a default certificate that will be
> served for all hostnames. In addition, `tls.routeOverride.certificate` is set to
> `Allow`, meaning Routes can specify TLS certificates for any domains. Next,
> there are two HTTPRoute resources which define certificates for
> `foo.example.com` and `bar.example.com`.

Reviewer: ```suggestion
In this example, the Gateway is configured with a default certificate that will be
served for all hostnames. In addition, `tls.routeOverride.certificate` is set to
`Allow`, meaning routes can specify TLS certificates for any domains. Next,
there are two HTTPRoute resources which specify certificates for
`foo.example.com` and `bar.example.com`.
```
I think we should use lower-cased "routes" for the general concept, and only capitalize when referring to a specific kind (such as HTTPRoute).

```yaml
{% include 'tls-cert-in-route.yaml' %}
```

## Upstream TLS

Upstream TLS configuration applies to the connection between the Gateway
and Service.

There is only one way to configure upstream TLS: using the `BackendPolicy`
resource.

Please note that the TLS configuration is related to the Service or backend

> resource and not related to a specific Route resource.

Reviewer: ```suggestion
resource and not related to a specific route resource.
```

### Example

The following example shows how upstream TLS can be configured. We have
omitted downstream TLS configuration for simplicity. As noted before, it
doesn't matter how downstream TLS is configured for the specific Listener or
Route.

```yaml
{% include 'upstream-tls.yaml' %}
```

## Extensions

Both upstream and downstream TLS configs provide an `options` map to add

> additional TLS settings for implementation specific features.

Reviewer: ```suggestion
additional TLS settings for implementation-specific features.
```

## Client/Server and TLS

![overview](tls-overview.svg)

For Gateways, there are two connections involved:

- **downstream**: This is the connection between the client and the Gateway.
- **upstream**: This is the connection between the Gateway and backend resources
   specified by routes. These backend resources will usually be Services.
  
With Service APIs, TLS configuration of downstream and
upstream connections is managed independently.

Please note that in case of `Passthrough` TLS mode, no TLS settings take
effect as the TLS session from the client is NOT terminated at the Gateway.
The rest of the document assumes that TLS is being terminated at the Gateway,
which is the default setting.

## Downstream TLS

Downstream TLS settings are configured using Listeners at the Gateway level.

### Listener and TLS 

Listeners expose the TLS setting on a per domain or sub-domain basis.
TLS settings of a listener are applied to all domains that satisfy the
`hostname.match` criteria.

In the following example, the Gateway serves the TLS certificate
defined in `default-cert` Secret resource for all requests.
Although, the example refers to HTTPS protocol, one can also use the same

> feature for TLS-only protocol alongwith TLSRoutes.

Reviewer 3: ```suggestion
feature for TLS-only protocol along with TLSRoutes.
```

# TLS details

Service APIs allow for a variety of ways to configure TLS. This document lays
out various TLS settings and gives general guidelines on how to use them
effectively.

## Client/Server and TLS

![overview](tls-overview.svg)

For Gateways, there are two connections involved:

- **downstream**: This is the connection between the client and the Gateway.
- **upstream**: This is the connection between the Gateway and backend resources
   specified by routes. These backend resources will usually be Services.
  
With Service APIs, TLS configuration of downstream and
upstream connections is managed independently.

Please note that in case of `Passthrough` TLS mode, no TLS settings take
effect as the TLS session from the client is NOT terminated at the Gateway.
The rest of the document assumes that TLS is being terminated at the Gateway,
which is the default setting.

## Downstream TLS

> Downstream TLS settings are configured using Listeners at the Gateway level.

Reviewer: ```suggestion
Downstream TLS settings are configured using listeners at the Gateway level.
```

> ### Listener and TLS 

Reviewer: ```suggestion
### Listeners and TLS 
```

Listeners expose the TLS setting on a per domain or sub-domain basis.
TLS settings of a listener are applied to all domains that satisfy the
`hostname.match` criteria.

In the following example, the Gateway serves the TLS certificate

> defined in `default-cert` Secret resource for all requests.

Reviewer: ```suggestion
defined in the `default-cert` Secret resource for all requests.
```

Although, the example refers to HTTPS protocol, one can also use the same
feature for TLS-only protocol along with TLSRoutes.

```yaml
listeners:
- protocol: HTTPS # Other possible value is `TLS`
  port: 443
  hostname:
    match: Any # Other possible values are `Domain` and `Exact`
  tls:
    mode: Terminate
    certificateRef:
      kind: Secret
      group: core
      name: default-cert
    routeOverride:
      certificate: Deny
```

If `hostname.match` is set to `Exact`, then the TLS settings apply to only the
specific hostname that is set in `hostname.name`.

Specifying `tls.routeOverride.certificate: Deny` is recommended because it
centralizes TLS configuration within the Gateway specification and should
suffice for the majority of use-cases. Please take a look at the examples below
for various alternatives.

> ### Route and TLS

Reviewer: ```suggestion
### Routes and TLS
```

If `listeners[].tls.routeOverride.certificate` is set to `Allow`, TLS certificates

> can be configured on Routes that are bound to the Gateway. This feature is

Reviewer: ```suggestion
can be configured on routes that are bound to the Gateway. This feature is
```

primarily meant for a cluster with a self-service model where Application developers
bring their own TLS certificates. This feature also mirrors the behavior of

> TLS as defined in the Ingress v1 resource. One should use this feature only
> when the Cluster Operator wishes to delegate TLS to Service or Application Developer
> the certificate defined in the Route will override any certificate defined in
> the Gateway

Reviewer: Looks like this sentence got corrupted?
```suggestion
TLS as defined in the Ingress v1 resource. One should use this feature only
when the Cluster Operator wishes to delegate TLS configuration to the Application Developer.
With this feature, the certificate defined in the route overrides any certificate defined in
the Gateway.
```

Since a specific match takes priority, the Gateway will serve
`foo-example-com-cert` for requests to `foo.example.com` and
`wildcard-example-com-cert` for all other requests.

```yaml
{% include 'wildcard-tls-gateway.yaml' %}
```

#### TLS Certificate in Route

In this example, the Gateway is configured with a default certificate that will be
served for all hostnames. In addition, `tls.routeOverride.certificate` is set to
`Allow`, meaning routes can specify TLS certificates for any domains. Next,
there are two HTTPRoute resources which specify certificates for
`foo.example.com` and `bar.example.com`.

```yaml
{% include 'tls-cert-in-route.yaml' %}
```

## Upstream TLS

Upstream TLS configuration applies to the connection between the Gateway
and Service.

There is only one way to configure upstream TLS: using the `BackendPolicy`
resource.

Please note that the TLS configuration is related to the Service or backend
resource and not related to a specific route resource.

### Example

The following example shows how upstream TLS can be configured. We have
omitted downstream TLS configuration for simplicity. As noted before, it

> doesn't matter how downstream TLS is configured for the specific Listener or
> Route.

Reviewer: ```suggestion
doesn't matter how downstream TLS is configured for the specific listener or
route.
```

# TLS

Service APIs allow for a variety of ways to configure TLS. This document lays
out various TLS settings and gives general guidelines on how to use them
effectively.

## Client/Server and TLS

![overview](tls-overview.svg)

For Gateways, there are two connections involved:

- **Downstream**: This is the connection between the client and the Gateway.

> - **Upstream**: This is the connection between the Gateway and Service. The
>   Service in question can be a core Kubernetes Service or any other network
>   endpoint as well.

Reviewer 4: This is really hard to describe well, I've never found a great way to say it, but maybe something like this?
```suggestion
- **Upstream**: This is the connection between the Gateway and backend resources
   specified by routes. These backend resources will usually be Services.
```

This is the recommended way of configuring TLS settings in Service APIs
as it does not involve any redirection and should suffice in majority of the
use-cases.

For more clarity, please take a look at the examples below.

### Route and TLS

If `listener.tls.routeOverride.certificate` is set to `Allow`, TLS certificates
can be configured at the Route-level.
This feature is primarily  meant for a cluster with a self-service model where
Service owners bring their own TLS certificates. This feature also mirrors
the behavior of TLS as defined in the Ingress v1 resource.
One should use this feature only when the Gateway Admin wishes to delegate
TLS to Service or App owners.

When using this feature, please note that TLS certificate to serve is chosen
before an HTTPRoute is selected. This is because TLS handshake is completed
before an HTTP request is sent from the client.

[TLS Certificate in Route](#tls-certificate-in-route) provides an example
on how this feature can be used.

### Examples

#### TLS in listener

In this example, Gateway is configured to serve `foo.example.com` and
`bar.example.com` domains. The certificate for these domains are specified
in Gateway.

> ```yaml

Author: That is a fair concern but I deleted the YAML files from the `examples` directory to ensure that I'm not increasing any duplication. I think it is better to include text with examples than having examples in a directory full of YAML files. I, as a user, mostly get lost in that.

Is there any upside of having these in YAML files and then referencing them? I guess one upside is that users can grab the URL and `kubectl apply` it.

> ```yaml

Reviewer 4: I think the biggest upside is that we'd have a way to automatically verify that our examples weren't stale.

> ```yaml

Author: Good point. Let me try to see how to get this to work in a separate PR (and hold this until that). I've no experience with mkdocs.

> ```yaml

Author: https://github.com/kubernetes-sigs/service-apis/pull/393

> ```yaml

Author: I've updated the PR with YAMLs inside `examples` and using file embedding approach.

#394 fixes a bug and should go in before this.

kind: Gateway
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: gateway
spec:
  gatewayClassName: acme-lb
  listeners:
  - protocol: HTTPS
    port: 443
    hostname:
      name: foo.example.com
      match: Exact
    tls:
      certificateRef:
        kind: Secret
        group: core
        name: foo-example-com-cert
      routeOverride:

>         certificate: Deny

Author: I considered this but decided against it because of two reasons:
- makes it clear that TLS config is ONLY in Listener
- for anybody trying to learn the TLS section of this API, it helps with keeping all bits in front. I've already remoed route selection section. 

That being said, I can reason either ways and don't feel strongly about it.
